flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/BroadcastRecordWriter.java

```diff
@@ -44,4 +52,67 @@ public BroadcastRecordWriter(
 	public void emit(T record) throws IOException, InterruptedException {
 		broadcastEmit(record);
 	}
+
+	@Override
+	public void broadcastEmit(T record) throws IOException, InterruptedException {
```

Comment:
I see couple of issues/potential issues here:
1. I'm not sure if I like the `randomTriggered` state field and how it's value affects the logic in different methods. It took quite some time to understand two successive `randomEmit` one after another work correctly (and how they actually work), because `requestNewBufferBuilder` behaves differently for `randomTriggered = true`. 
2. I'm not sure, but I think that `BroadcastRecordWriter`'s logic would be actually easier to follow, if it didn't have `bufferBuilders` array, but single `bufferBuilder` field. Maybe it would be worth creating `AbstractRecordWriter` without `bufferBuilders`/`bufferBuilder` field and then have two concrete implementations of `RecordWriter` and `BroadcastRecordWriter` with one two different implementations?
3. `LatencyMarker`s are a bit broken here, since effectively they will be flushed on the very first next emit following the `LatencyMarker`, so they do not measure the actual latency. This might not be the critical issue, but it would be good to have correct metric for that.

Point 1. is not that difficult to address (for 1. `randomEmit()` could flush the current buffer, request a new one for one `randomEmit()` and immediately finish it, so that this special handling of `randomTriggered = true` wouldn't leak outside of the `randomEmit()` method).

Point 2. Might be not worth doing so.

Point 3. is the more problematic one I guess. 

a) One way to deal with it, would be to broadcast `LatencyMarker`. Easy to do, but with huge number of output channels and multiple broadcasts channels one after another, would result with gigantic cascade/avalanche of the `LatencyMarkers`. 
b) Another idea would be to broadcast the `LatencyMarker` to everybody, but add some kind of marker to let the receiver know, that it should ignore it, because it was targeted to someone else.
c) Maybe there are other solutions that I haven't thought about?

Reply:
Thanks for these kindly detail suggestions. You pointed out a very critical potential problem which I have not thought through before.

Regarding with the `LatencyMarker`, I have not thought of a perfect solution yet.
1. If finishing the `LatencyMarker` immediately, it could only reflect the latency before this marker, and lose a bit latency of current `BufferConsumer` containing `LatencyMarker`.

2. If broadcasting the `LatencyMarker`, there are two concerns: First we might add the `LatencyMarkerID` in the marker like `CheckpointID`, and ignore the duplicate id in `StreamInputProcessor` which needs maintain a set of `LatencyMarkerID` for filter.  We could add this logic in processor to avoid gigantic avalanche of the marker. If to do so, the logic seems very easy for `BroadcastRecordWriter`, and the methods for `emit` and `randomEmit` are all processed in broadcast mode. But if we extend other elements to be `randomEmit` future, we might need add corresponding special handing logic like `LatencyMarker` then. The second concern is that broadcasting `LatencyMarker` still loses the latency precision compared with now. Currently we choose the random channel to send this marker, so it could reflect the average latency from statistics. But the broadcast mode could always reflect the latency of the fastest channel, because once the downstream receives the first marker, it would ignore the same following markers from other slow channels.

Considering the above two options, I would prefer the first option because the latency lose seems less compared with broadcasting way, and we do not need to adjust the logic on receiver side (how many IDs to be maintained in memory might be another concern in this way). 

Regarding with the first option, I agree with your suggestion of abstracting `AbstractRecordWriter` to maintain single `BufferBuilder` in `BroadcastRecordWriter`.  And  I could understand your concerns of the `randomTriggered` field. Your suggestion of finishing `LatencyMarker` immediately in `randomEmit` could avoid this field. I would like to have a try based on them if you also approve the first option. :)

Reply:
Thought of another two points:
- `randomTriggered` field could not be removed completely, because the behaviors are different in the methods of `requestNewBufferBuilder` and `flushTargetPartition` based on this tag. Otherwise the current `copyFromSerializerToTargetChannel` could not be reused and `BroadcastRecordWriter` has to implement the private process of `randomEmit`. If we could retain this tag field, and finish current `BufferBuilder` immediately after `randomEmit`. Then only `randomEmit` would change this field, and only `requestNewBufferBuilder` and `flushTargetPartition` would read this tag. It would seem easy to understand compared with now.

- Considering immediately finishing current `BufferBuilder` once `randomEmit` done, it might increase the frequency of finishing `BufferBuilder`. For example, if the `randomEmit` is triggered very frequently and there are no common `emit` or `broadcastEmit` calling for a while, one `BufferBuilder` might write multiple random emitted elements, no need one `BufferBuilder` for one `randomEmit`. Or this worry is not worth considering for most scenarios?

Reply:
I also do not like spilling this logic to the receiver. I thought a little bit more about it and there is also one more option that should work. 

We have one shared `BufferBuilder` with multiple `BufferConsumers`. Once we get `randomEmit()` for a single channel `channelId`, we:
1. finish the current `BufferBuilder`
2. we create a new `BufferBuilder` (let's call it `bb`)
3. we serialize `LatencyMarker` to the `bb`
4. we enqueue new `BufferConsumer` created from `bb` to the `channelId`
5. For the sake of other channels, we create `BufferConsumer`s from the `bb` with offsetted starting position. Starting position should be offseted in a such way, that the `LatencyMarker` wouldn't be visible for those channels. (initiating `BufferConsumer#currentReaderPosition` to `LatencyMarker.size()`).

There would be one drawback of this solution. Reported latency, would be slightly higher, since latency markers would be always written to the buffer's beginning, thus marker would have to wait a bit more for the buffer to fill in compared to how long the regular records do wait on average. However flushing timeout wouldn't be affected, so in the end this should be fine.

What do you think?

Reply:
Thanks for thinking of one more possible solution for this issue. I actually like the idea of sharing random `BufferBuilder` in all partitions by setting the offset position of `BufferConsumer`. In the before I also thought of another similar way of always sharing `BufferBuilder` no matter with broadcast switch, then every `BufferConsumer` maintains the different skip list to filter the data not belonging to itself. But it seems so complicated and might bring extra overheads. Your suggestion make it easy to only skip fixed length at the beginning of `BufferConsumer`.

I am not so worried about the latency higher as you mentioned because we also have flushing mechanism to work around. But I am just confused of another issue, that means we strongly couple  the `randomEmit` with `LatencyMarker` together. The`randomEmit` should be a common method provided for operators, but the implementation seems not extendable,  even though it might never have other usages for `randomEmit` in future.

Reply:
Regarding with abstracting `RecordWriter`, I have some plans for your confirm.

1. Make current `RecordWriter` as abstract class, so it is no need to refactor existing references in other places. Then it has two different implementations `BroadcastRecordWriter` and `SelectorRecordWriter` separately.

2.  There are two main differences for these two implementations. `BroadcastRecordWriter` maintain single `BufferBuilder` and no need `ChannelSelector`. `SelectorRecordWriter` maintains the array of `BufferBuilder` and has the `ChannelSelector`.

3. Based on above two differences, we might need to abstract related methods such as `emit`, `randomEmit`, `broadcastEmit`, `getBufferBuilder`, `tryFinishCurrentBufferBuilder`, `requestNewBufferBuilder`, `closeBufferBuilder`.

4. In order to reuse the common main process of `RecordWriter#copyFromSerializerToTargetChannel` for all kinds of emit, we still need the boolean flag `randomTriggered` in `BroadcastRecordWriter`, because it would have two different paths in `requestNewBufferBuilder` and `flushTargetPartition` based on this condition. And this flag would only be updated to true at the beginning of `BroadcastRecordWriter#randomEmit`, and updated to false at the end of this method.

Do you have other concerns for above thoughts ?

Reply:
I don't think that we strongly couple `randomEmit` and `LatencyMarker` with my last proposal. The only "bad" thing that I see which we would be doing is that `randomEmit()` will have a performance penalties if used on `BroadcastRecordWriter`, which I guess is an acceptable trade off. If we start using `randomEmit()` for more use cases, we might just need to make selection of regular vs `BroadcastRecordWriter` either smarter or configurable.

points 1 to 3 make sense to me. I don't fully understand the issue with `copyFromSerializerToTargetChannel`, but this is a small change, so it might be just better if you write the code and show it in the PR instead of explaining.

Reply:
Considering initiating `BufferConsumer#currentReaderPosition` to `LatencyMarker.size()` for other channels in `BroadcastRecordWriter#randomEmit`, I meant this way of getting marker size seems coupled with `LatencyMarker` currently. Maybe we do not need consider it now, if there are other usages in future, we can make it smarter then.

For `copyFromSerializerToTargetChannel`, this process might be called either by `randomEmit` or `BroadcastRecordWriter#emit`. And the `requestNewBufferBuilder` would be triggered in `copyFromSerializerToTargetChannel`. If it is in the `randomEmit` process, we need to initialize `BufferConsumer#currentReaderPosition` for other sharing channels, otherwise we do not need to do. So it needs a flag to distinguish the different ways.  It would be easy for explanation after I submit the codes later. :)

Reply:
> Considering initiating BufferConsumer#currentReaderPosition to LatencyMarker.size() for other channels in BroadcastRecordWriter#randomEmit, I meant this way of getting marker size seems coupled with LatencyMarker currently. 

We don't have to hardcode `LatencyMarker.size()`. After all when we were about to set the offset, we have just serialized the record for `randomEmit()` to a fresh `BufferBuilder` :)

Reply:
Yes, you are right. We do not need hardcode the size of latency marker. :)

I have submitted the codes in a separate commit. I am not sure whether this changes is suitable based on our comments, so I only refactored the main processes and ignored the tests with annotations. After your review and confirm, I would refactor the commits and tests.

I think it is better to explain some thoughts below:

In order to reuse the main process of `RecordWriter#copyFromSerializerToTargetChannel`, the `randomTriggered` field is still maintained in `BroadcastRecordWriter`, but I think the logic is easy to understand than the first version.

After abstracting the `RecordWriter` to separate two implementations, the changes seem more complex, because there are many abstract methods involved in `BufferBuilder` and `ChannelSelector` which are the main differences between two implementations. 

I am not sure you could accept the way of magic 0 in `BroadcastRecordWriter#broadcastEmit` for reusing the common codes. If we do not pass the target channel in broadcast operation, we might need to realize all the related processes in `BroadcastRecordWriter`.

I think my first version seems more easy to follow if we only change the process of `BroadcastRecordWriter#randomEmit` as now. Maybe you have other good ways. :)

Reply:
Ok, I see the problem now with `randomTriggered`. I don't like it, but also I don't have anything better to suggest :( So let it be.

> After abstracting the RecordWriter to separate two implementations, the changes seem more complex

true. The alternate solution would be just dump the base class, but I don't think that it would be better. 

> I am not sure you could accept the way of magic 0

I noticed it before and again I'm not a big fan of it, but I can not find a better solution.

> think my first version seems more easy to follow

I'm not sure. It had less code/changes. Issue is that in that previous version, `BroadcastRecordWriter` was still using array of `BufferBuilder`s and this is also confusing and probably less performant. If you could run network benchmarks showing that both versions are equally performant I would be ok-ish with either one.

I was expecting this to be a simple couple of lines change, but clearly I have missed judged it :(

Reply:
The fist version is just trying best to reuse the common codes. But it seems not good to understand and distinguish these two writers in essential, including `BufferBuilder` and `ChannelSelector`. Furthermore in theory the single `BufferBuilder` seems more efficient for `BroadcastRecordWriter`. From this point, I am willing to keep the current version if there are no better solutions currently. 

I would refactor the commit and related tests based on current idea if you have no other concerns. :)

Reply:
Sounds good :)